Fix issue with PostgreSQL: org.postgresql.util.PSQLException

nenaraab opened the pull request:
When trying to use Spring Security ACL, I am facing the error: org.postgresql.util.PSQLException: ERROR: operator does not exist: bigint = character varying.

As recommended in the [Spring.io documentation](https://docs.spring.io/spring-security/site/docs/current/reference/html5/#postgresql) i've setup the acl tables in my PostgreSQL database, with column `acl_object_identity.object_id_identity` of type `varchar(36)`.

IMHO the `object_id_identity` in the `jdbcTemplate` method calls MUST always be passed as a character sequence.

This relates to issue:
#5508

pivotal-issuemaster commented:
@nenaraab Please sign the [Contributor License Agreement](https://cla.pivotal.io/sign/spring?repositoryId=spring-projects/spring-security&pullRequestId=6050)!

[Click here](https://cla.pivotal.io/sync/spring?repositoryId=spring-projects/spring-security&pullRequestId=6050) to manually synchronize the status of this Pull Request.

See the [FAQ](https://cla.pivotal.io/about) for frequently asked questions.

pivotal-issuemaster commented:
@nenaraab Thank you for signing the [Contributor License Agreement](https://cla.pivotal.io/sign/spring?repositoryId=spring-projects/spring-security&pullRequestId=6050)!

mangei commented on acl/src/main/java/org/springframework/security/acls/jdbc/JdbcAclService.java:
```diff
@@ -88,7 +88,7 @@ public JdbcAclService(DataSource dataSource, LookupStrategy lookupStrategy) {
 
 	public List<ObjectIdentity> findChildren(ObjectIdentity parentIdentity) {
-		Object[] args = { parentIdentity.getIdentifier(), parentIdentity.getType() };
+		Object[] args = { "" + parentIdentity.getIdentifier(), parentIdentity.getType() };
```
What about calling `toString()` instead of doing string concatination? Or may the objects be `null`?

```suggestion
		Object[] args = { parentIdentity.getIdentifier().toString(), parentIdentity.getType() };
```

nenaraab replied:
@mangei Thanks! 
No the identifier could never be null, I've adapted the code accordingly!

rwinch commented:
Thank you for the PR @nenaraab! 

1) Could you please rebase off master?
2) Could you please use `String.valueOf` as suggested by @mangei?

If you need any help with the rebase, please let me know and I'd be happy to help.

nenaraab commented:
@rwinch 
Thanks for cooperation! I've updated the PR. This will fix all my current PostgreSQL issues...

rwinch commented on acl/src/main/java/org/springframework/security/acls/jdbc/JdbcAclService.java:
```diff
@@ -88,7 +88,7 @@ public JdbcAclService(DataSource dataSource, LookupStrategy lookupStrategy) {
 
 	public List<ObjectIdentity> findChildren(ObjectIdentity parentIdentity) {
-		Object[] args = { parentIdentity.getIdentifier(), parentIdentity.getType() };
+		Object[] args = { parentIdentity.getIdentifier().toString(), parentIdentity.getType() };
```
Can you please account for the fact that this `getIdentifier()` might be `null`? Initially I thought this would be best with `String.valueOf`, but that converts `null` to `"null"` which is not ideal. Instead I think we should do a `null` check and if null use `null`.

NOTE: Rather than commenting in each spot, can you please apply this change throughout

nenaraab commented:
@rwinch
no problem! Please confirm... i will add some checks (using `Assert`) at the right places to make it bulletproof
- check that identifier is not NULL (if not already checked in the calling public method)
- check that `Spring.valueOf` does not return null

rwinch commented:
Thanks for the fast reply @nenaraab!

Sorry for the confusion. I was not very clear. What I meant was to do something like this:

```java
String id = parentIdentity.getIdentifier() == null ? 
    null : parentIdentity.getIdentifier().toString()
```

As for the testing I meant to add some tests that break before your changes and are fixed after the changes. Also add a test that validates that a null value is passed through if `getIdentifier()` is passed in. I'd probably add them to [JdbcAclServiceTests](https://github.com/spring-projects/spring-security/blob/master/acl/src/test/java/org/springframework/security/acls/jdbc/JdbcAclServiceTests.java#L54).

To make testing more realistic, I'd:
* [Change JdbcTemplate](https://github.com/spring-projects/spring-security/blob/5.1.2.RELEASE/acl/src/main/java/org/springframework/security/acls/jdbc/JdbcAclService.java#L69) to be `JdbcOperations`
* Add a constructor that allows injecting the `JdbcOperations` and `LookupStrategy`

Then the tests could inject a Mock `JdbcOperations` for the tests mentioned above.

Thanks again for your help with this ticket! If you have any questions, don't hesitate to reach out.

nenaraab commented:
@rwinch 
It's not that i dislike to implement a unittest... and thanks for the guidance here.

But as `ObjectIdentity.getIdentifier()` never returns `null` (see [documentation](https://github.com/spring-projects/spring-security/blob/5.1.2.RELEASE/acl/src/main/java/org/springframework/security/acls/model/ObjectIdentity.java#L57), see [all constructor implementations](https://github.com/spring-projects/spring-security/blob/5.1.2.RELEASE/acl/src/main/java/org/springframework/security/acls/domain/ObjectIdentityImpl.java#L45)) i don't get the point of introducing a null check at all.

Furthermore I've checked the ObjectIdentityImplTests thoroughly. All not null assertions according to the identifier are covered by tests.

rwinch commented on acl/src/test/java/org/springframework/security/acls/jdbc/JdbcAclServiceTests.java:
```diff
@@ -57,8 +57,8 @@ public void setUp() {
 	public void readAclByIdMissingAcl() {
 		Map<ObjectIdentity, Acl> result = new HashMap<>();
 		when(
-				lookupStrategy.readAclsById(anyListOf(ObjectIdentity.class),
```
If we modify existing tests we run the risk of breaking other functionality later down the road. Please leave existing tests as is and create a new test that validates the new beahvior.

nenaraab replied:
Sorry, my fault, I was not clear on the change of the test case.

I've created a new issue / pull request to discuss and fix that issue:
https://github.com/spring-projects/spring-security/issues/6212

And I've have reverted the change on the test case :-)

rwinch replied:
No problem. Thanks for reverting that and submitting the other PR.

Can you still add [some tests](https://github.com/spring-projects/spring-security/pull/6050#issuecomment-443025708) please? No need to add one for the null value (which as you pointed out cannot happen), but we should add one for the issue that we are fixing (i.e. ensure that non String types are converted).

Thanks!

nenaraab commented:
@rwinch 
great, added two test cases to test the code change.

Furthermore I've added integration tests (using hsql in-memory database) to test also a proper conversion of the fetched entries, and the find-children sql queries as well.

nenaraab commented on acl/src/main/java/org/springframework/security/acls/jdbc/JdbcAclService.java:
```diff
@@ -79,7 +80,16 @@
+		this.aclClassIdUtils = new AclClassIdUtils();
+	}
+
+	public JdbcAclService(DataSource dataSource, LookupStrategy lookupStrategy, JdbcOperations jdbcOperations) {
```
Great, like your suggestion! I've kept the `Assert.notNull(dataSource, "DataSource required");` in the "old" constructor for compatibility reason...

rwinch commented on acl/src/main/java/org/springframework/security/acls/jdbc/JdbcAclService.java:
```diff
@@ -77,9 +78,14 @@
 
 	public JdbcAclService(DataSource dataSource, LookupStrategy lookupStrategy) {
+		this(new JdbcTemplate(dataSource), lookupStrategy);
 		Assert.notNull(dataSource, "DataSource required");
```
this assertion is not necessary because JdbcTemplate validates that the dataSource is not null.

nenaraab replied:
I'm so sorry! 
[this @Nullable`](https://github.com/spring-projects/spring-framework/blob/master/spring-jdbc/src/main/java/org/springframework/jdbc/support/JdbcAccessor.java#L45) was missleading...
but you're right: the method `afterPropertiesSet` raises an exception, so I've removed the assertion. Sorry for inconvenience...

rwinch replied:
No problem. It is a bit confusing and due the fact there is also a default constructor.

Thanks for the fast turnaround!

juzerali commented:
In which version is this fix available?

jzheaux commented:
You can see by looking at the commit hash - https://github.com/spring-projects/spring-security/commit/d1a754fcf25ef0e59b0c46f80f34b0921ecd1c13. In the comment at the top, the versions are listed. Let me know if I can help further.

juzerali commented:
Got it, so it is available since v5.2. Thanks @jzheaux
